js/piwik.js

```diff
@@ -7934,8 +7966,39 @@ if (typeof window.Piwik !== 'object') {
                 for (i; i < missedCalls.length; i++) {
                     apply(missedCalls[i]);
                 }
+            },
+
+            /**
+             * Send a message to the iframe which holds the opt-out form, telling it the current value of the 
+             * first-party opt-out cookie.
+             */
+            updateOptOutForm: function() {
+                if (typeof window.postMessage === 'undefined') {
+                    // We're on an older browser so we can't do anything
+                    return;
+                }
+
+                // We don't know whether iframe has finished loading yet so we'll fire it a few times per second
+                // The iframe will send a response back to us which we handle in our message listener by
+                // clearing this interval.
+                var optOutStatus = {opted_in: this.getTracker().hasConsent()};
+                var numAttempts = 0;
+                optOutTimer = setInterval(function() {
+                    var iframes = document.getElementsByTagName('iframe');
+                    for (var i = 0; i < iframes.length; i++) {
+                        var iframe = iframes[i];
+                        iframe.contentWindow.postMessage(JSON.stringify(optOutStatus), '*');
+                    }
+                    numAttempts++;
+                    // 10 times per second * 1200 = 2 minutes
+                    // If the iframe hasn't finished loading by now, it ain't gonna, so let's stop trying
+                    if (numAttempts > 1200) {
+                        clearInterval(optOutTimer);
+                    }
+                }, 100);
```

Comment:
As DOM interactions are quite expensive in terms of performance, be good to check this maybe only about every second or so?

Reply:
The problem is that the optout iframe might be displaying incorrectly to the user (e.g. saying they haven't opted out when they have) until it receives this message.  The first time through the loop is usually unsuccessful because the iframe hasn't finished loading yet, and increasing the delay even to 200ms means there is a noticeable pause before the optout iframe updates itself.  Maybe we need a new approach - calling `postMessage` on the iframe before it finishes loading is also causing an error message to be printed to the console which isn't ideal.

Reply:
Could use `setTimeout` instead... and each time use something like `interval = numAttempts * 100`